model/config.go

```diff
@@ -24,75 +23,79 @@ import (
-}
-
-// String returns a human-readable textual representation of a key
-func (k Key) String() string {
-	return fmt.Sprintf("%s/%s-%s", k.Namespace, k.Kind, k.Name)
-}
-
-// ConfigRegistry defines the basic API for retrieving and storing
-// configuration artifacts.
+// The storage interface presented here assumes that the underlying storage
+// layer supports _GET_ (list), _PUT_ (update), _POST_ (create) and _DELETE_
+// semantics but does not guarantee any transactional semantics.
+//
+// Configuration objects possess a type property (also called kind throughout)
+// and a key property. The configuration key is derived from the content of the
+// config object and uniquely identifies objects for a particular kind. For
+// example, if the object schema contains metadata fields _name_ and
+// _namespace_, the configuration key may be defined as _namespace/name_.
+// Key definition is provided as part of the config kind definition.
 //
-// "Put", "Post", and "Delete" are mutator operations. These operations are
-// asynchronous, and you might not see the effect immediately (e.g. "Get" might
+// _PUT_, _POST_, and _DELETE_ are mutator operations. These operations are
+// asynchronous, and you might not see the effect immediately (e.g. _GET_ might
 // not return the object by key immediately after you mutate the store.)
 // Intermittent errors might occur even though the operation succeeds, so you
 // should always check if the object store has been modified even if the
-// mutating operation returns an error.
+// mutating operation returns an error.  Objects should be created with _POST_
+// operation and updated with _PUT_ operation.
 //
-// Objects should be created with "Post" operation and updated with "Put" operation.
+// Revisions record the last mutation operation on each object. If a mutation
+// is applied to a different revision of an object than what the underlying
+// storage expects as defined by pure equality, the operation is blocked.  The
+// client of this interface should not make assumptions about the structure or
+// ordering of the revision identifier.
 //
 // Object references supplied and returned from this interface should be
 // treated as read-only. Modifying them violates thread-safety.
-type ConfigRegistry interface {
-	// Get retrieves a configuration element, bool indicates existence
-	Get(key Key) (proto.Message, bool)
+type ConfigStore interface {
+	// KindMap exposes the configuration type schema known by the config store.
+	// The type schema defines the bidrectional mapping between configuration
+	// types (also called kinds) and the protobuf message names.
+	KindMap() KindMap
 
-	// List returns objects for a kind in a namespace.
-	// Use namespace "" to list resources across namespaces
-	List(kind string, namespace string) (map[Key]proto.Message, error)
+	// Get retrieves a configuration element by a kind and a key
+	Get(kind, key string) (config proto.Message, exists bool, revision Revision)
```

Comment:
The key uniquely identifies a config. The structured keys are not part of this design. The two can be unified as: `List(kind string, keyPrefix []string) ([]Config, error)`